src/collector.rs

```diff
@@ -1,63 +1,282 @@
 use crate::extract::{extract_links, FileType};
 use crate::types::Uri;
-use anyhow::Result;
-use glob::glob;
+use anyhow::{anyhow, Context, Result};
+use glob::glob_with;
 use reqwest::Url;
-use std::{collections::HashSet, fs};
-use std::{ffi::OsStr, path::Path};
+use shellexpand::tilde;
+use std::collections::HashSet;
+use std::path::Path;
+use std::path::PathBuf;
+use tokio::fs::read_to_string;
+use tokio::io::{stdin, AsyncReadExt};
 
-/// Detect if the given path points to a Markdown, HTML, or plaintext file.
-fn resolve_file_type_by_path<P: AsRef<Path>>(path: P) -> FileType {
-    match path.as_ref().extension().and_then(OsStr::to_str) {
-        Some("md") => FileType::Markdown,
-        Some("html") => FileType::HTML,
-        _ => FileType::Plaintext,
+const STDIN: &str = "-";
+
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub(crate) enum Input {
+    RemoteUrl(Url),
+    FsGlob { pattern: String, ignore_case: bool },
+    FsPath(PathBuf),
+    Stdin,
+    String(String),
+}
+
+#[derive(Debug)]
+pub(crate) struct InputContent {
+    pub input: Input,
+    pub file_type: FileType,
+    pub content: String,
+}
+
+impl InputContent {
+    pub fn from_string(s: &str, file_type: FileType) -> Self {
+        // TODO: consider using Cow (to avoid one .clone() for String types)
+        Self {
+            input: Input::String(s.to_owned()),
+            file_type,
+            content: s.to_owned(),
+        }
+    }
+}
+
+impl Input {
+    pub(crate) fn new(value: &str, glob_ignore_case: bool) -> Self {
+        if value == STDIN {
+            Self::Stdin
+        } else {
+            match Url::parse(&value) {
+                Ok(url) => Self::RemoteUrl(url),
+                Err(_) => {
+                    // this seems to be the only way to determine if this is a glob pattern
+                    let is_glob = glob::Pattern::escape(value) != value;
+
+                    if is_glob {
+                        Self::FsGlob {
+                            pattern: value.to_owned(),
+                            ignore_case: glob_ignore_case,
+                        }
+                    } else {
+                        Self::FsPath(value.into())
+                    }
+                }
+            }
+        }
+    }
+
+    pub async fn get_contents(
+        &self,
+        file_type_hint: Option<FileType>,
+        skip_missing: bool,
+    ) -> Result<Vec<InputContent>> {
+        use Input::*;
+
+        match self {
+            // TODO: should skip_missing also affect URLs?
```

Comment:
Good question and no idea. Let's leave that edge-case open for now. Pretty sure we'll eventually know the right answer. 😉